python/pyspark/mllib/linalg/__init__.py

```diff
@@ -461,32 +461,41 @@ def __init__(self, size, *args):
         self.size = int(size)
         """ Size of the vector. """
         assert 1 <= len(args) <= 2, "must pass either 2 or 3 arguments"
-        if len(args) == 1:
-            pairs = args[0]
-            if type(pairs) == dict:
-                pairs = pairs.items()
-            pairs = sorted(pairs)
-            self.indices = np.array([p[0] for p in pairs], dtype=np.int32)
-            """ A list of indices corresponding to active entries. """
-            self.values = np.array([p[1] for p in pairs], dtype=np.float64)
-            """ A list of values corresponding to active entries. """
+        if isinstance(args[0], bytes):
+            assert isinstance(args[1], bytes), "values should be string too"
+            if args[0]:
+                self.indices = np.frombuffer(args[0], np.int32)
+                self.values = np.frombuffer(args[1], np.float64)
+            else:
+                # np.frombuffer() doesn't work well with empty string in older version
+                self.indices = np.array([], dtype=np.int32)
+                self.values = np.array([], dtype=np.float64)
         else:
-            if isinstance(args[0], bytes):
-                assert isinstance(args[1], bytes), "values should be string too"
-                if args[0]:
-                    self.indices = np.frombuffer(args[0], np.int32)
-                    self.values = np.frombuffer(args[1], np.float64)
-                else:
-                    # np.frombuffer() doesn't work well with empty string in older version
-                    self.indices = np.array([], dtype=np.int32)
-                    self.values = np.array([], dtype=np.float64)
+            if len(args) == 1:
+                args = args[0]
+                if isinstance(args, dict):
+                    args = args.items()
+                args = list(zip(*args))
+
+            # Handle empty args case.
+            if len(args) == 0:
```

Comment:
I was thinking something more along the lines of

```python
        if len(args) == 1:
            args = args[0]

            # Handle empty args case.
            if len(args) == 0:
                args = [[], []]
            elif isinstance(args, dict):
                args = args.items()
            args = list(zip(*args))
        indices, values = args
```

You don't have to worry about `len(args) == 0` because of L463 and this makes it so `len(args) == 0` is only checked when args can possibly be `[[]]` or `[dict()]` (the current check will also check when `args = [[0,1,2],[3,4,5]]`)